### PR TITLE
Add systemd-resolved mDNS advertiser support

### PR DESCRIPTION
**Pull request by elyscape**
## :recycle: Current situation

Homebridge currently supports [three different mDNS advertisers][mdns], two that run in-process and one that uses Avahi over D-Bus. For a while now, systemd-resolved has supported acting as an mDNS advertiser and resolver. If Homebridge is run on a system that uses systemd-resolved for mDNS, this can cause conflicts similar to homebridge/homebridge#3077.

## :bulb: Proposed solution

systemd-resolved, like Avahi, exposes [a D-Bus interface][dbus] for managing mDNS services. This pull request adds a systemd-resolved mDNS advertiser that makes use of this interface. The code was heavily based on that of the Avahi advertiser.

## :gear: Release Notes

- Added a new mDNS advertiser named `resolved` that uses systemd-resolved via D-Bus.

## :heavy_plus_sign: Additional Information

In order for homebridge to have the permission to use this interface, a `.pkla` file with the following contents will need to be installed to `/etc/polkit-1/localauthority/10-vendor.d/`:
```
[Homebridge RegisterService]
Identity=unix-user:homebridge
Action=org.freedesktop.resolve1.register-service
ResultAny=yes

[Homebridge UnregisterService]
Identity=unix-user:homebridge
Action=org.freedesktop.resolve1.unregister-service
ResultAny=yes
```

### Testing

Seems to work for me!

See: homebridge/homebridge#3224

[mdns]: https://github.com/homebridge/homebridge/wiki/mDNS-Options
[dbus]: https://www.freedesktop.org/software/systemd/man/org.freedesktop.resolve1.html

### CODE REVIEW

**Comment by elyscape** [comment]
Dealt with conflicts from master and ESLint warning. This should be ready whenever.

File: src/lib/Advertiser.ts [hunk 1]
```diff
@@ -422,3 +422,171 @@ export class AvahiAdvertiser extends EventEmitter implements Advertiser {
...
+
+  public async updateAdvertisement(silent?: boolean): Promise<void> {
+    if (!this.bus) {
+      throw new Error("Tried to update systemd-resolved advertisement on a destroyed advertiser!");
+    }
+
+    debug("Updating txt record (txt: %o, silent: %d)", CiaoAdvertiser.createTxt(this.accessoryInfo, this.setupHash), silent);
+
+    await this.stopAdvertising();
+    await this.startAdvertising();
```
Comment (Supereg) [inline src/lib/Advertiser.ts:493]: Is there no better way to update the txt record with `systems-resolved`? Readvertising the whole service is a pretty heavy operation and is actively discouraged for updates!

File: src/lib/Advertiser.ts [hunk 2]
```diff
@@ -422,3 +422,171 @@ export class AvahiAdvertiser extends EventEmitter implements Advertiser {
...
+    }
+
+    if (this.path) {
+      try {
+        await ResolvedAdvertiser.resolvedInvoke(this.bus, "UnregisterService", {
+          body: [this.path],
+          signature: "o",
+        });
+      } catch (error) {
+        // Typically, this fails if e.g. systemd-resolved service was stopped in the meantime.
```
Comment (Supereg) [inline src/lib/Advertiser.ts:508]: I'm currently implementing a mechanism to detect restarts of the avahi-daemon for the avahi-based dbus advertiser (see https://github.com/homebridge/HAP-NodeJS/pull/970). Could we pull off something similar with systems-resolved interface?

File: src/lib/Advertiser.ts [hunk 3]
```diff
@@ -422,3 +422,171 @@ export class AvahiAdvertiser extends EventEmitter implements Advertiser {
...
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      bus.invoke(command, (err: any, result: any) => {
+        if (err) {
+          reject(new Error(`resolvedInvoke error: ${JSON.stringify(err)}`));
+        } else {
+          resolve(result);
+        }
+      });
+    });
+  }
```
Comment (Supereg) [inline src/lib/Advertiser.ts:591]: This is just a plain copy of the code of the avahi-based advertiser right? Could we try to reuse the existing code a bit more?

**Review by Supereg — COMMENTED** [review]
Thanks for your contribution 🚀 
Had only a few comments on certain places ☺️ 

You mentioned, that there is some initial setup required to set permissions for homebridge/hap-nodejs. Can we please add this to the in-code documentation so users can easily find that.
Further, can programatically detect, if we are missing the permission for systemd-resolved such that we can issue an according warning (maybe with a link to a Wiki documentation page illustration how to configure permissions)?

File: src/lib/Advertiser.ts (hunk 1 above)
Comment (elyscape) [reply src/lib/Advertiser.ts:493]: Unfortunately, this is currently the case.

File: src/lib/Advertiser.ts (hunk 2 above)
Comment (elyscape) [reply src/lib/Advertiser.ts:508]: Possibly. I'd need some guidance on what your PR actually does, and then I'd need to test to see if it worked for systemd-resolved, but in theory it should be doable.

File: src/lib/Advertiser.ts (hunk 3 above)
Comment (elyscape) [reply src/lib/Advertiser.ts:591]: What would you recommend as the best way to do that?

**Comment by elyscape** [comment]
> Can we please add this to the in-code documentation so users can easily find that.

Sure. Where should I put it?

> Further, can programatically detect, if we are missing the permission for systemd-resolved such that we can issue an according warning (maybe with a link to a Wiki documentation page illustration how to configure permissions)?

If we're missing the appropriate permissions, there are a number of different errors that can occur as a result. Rather than try to figure out what they all are, how do you feel about making the error handler always issue that warning in response to failures?

File: src/lib/Advertiser.ts (hunk 1 above)
Comment (elyscape) [reply src/lib/Advertiser.ts:514]: I added a comment explaining this.

File: src/lib/Advertiser.ts (hunk 3 above)
Comment (elyscape) [reply src/lib/Advertiser.ts]: I extracted these methods into shared functions.

File: src/lib/Advertiser.ts (hunk 2 above)
Comment (elyscape) [reply src/lib/Advertiser.ts:529]: I did some digging and unfortunately systemd-resolved's D-Bus interface doesn't have any signals for us to listen onto. As far as I can tell, our best bet to detect restarts is to periodically poll for the existence of the object on the bus and re-register if it's missing.

**Comment by elyscape** [comment]
@Supereg Any thoughts on what else I should do to get this in a mergable state?

**Comment by elyscape** [comment]
Tests should work now, btw.

File: src/lib/Advertiser.ts (hunk 2 above)
Comment (Supereg) [reply src/lib/Advertiser.ts:524]: > I did some digging and unfortunately systemd-resolved's D-Bus interface doesn't have any signals for us to listen onto.

That's a pity. 

> As far as I can tell, our best bet to detect restarts is to periodically poll for the existence of the object on the bus and re-register if it's missing.

But also this seams like there is not an easy way for us to check this right? The only method that takes the `service_path` as an input is the `UnregisterService` method.
So I think the best is to just document that for now and treat that feature as an experimental feature as for now?

**Review by Supereg — APPROVED** [review]

**Comment by donavanbecker** [comment]
@Supereg, I have opened this Pull Request on the UI to add resolved: https://github.com/oznu/homebridge-config-ui-x/pull/1427

**Pull request merged**